fdbserver/DataDistribution.actor.cpp

```diff
@@ -1512,219 +1631,690 @@ ACTOR Future<Void> ddGetMetrics(GetDataDistributorMetricsRequest req,
+	try {
+		ServerWorkerInfos serverWorkers = wait(self->txnProcessor->getServerListAndProcessClasses());
+		int selected = deterministicRandom()->randomInt(0, serverWorkers.servers.size());
+		audit->actors.add(auditMakeProgressOnServer(
+		    self, audit, range, serverWorkers.servers[selected].first, /*makeProgressByServer=*/false));
+	} catch (Error& e) {
+		TraceEvent(SevWarn, "DDRunAuditJobBySingleServerError", self->ddId)
+		    .errorUnsuppressed(e)
+		    .detail("AuditID", audit->coreState.id)
+		    .detail("AuditType", audit->coreState.getType());
+		throw e;
+	}
+	TraceEvent(SevInfo, "DDRunAuditJobBySingleServerEnd", self->ddId)
+	    .detail("AuditID", audit->coreState.id)
+	    .detail("AuditType", audit->coreState.getType());
+
+	return Void();
+}
+
+// For each of storage servers, run an audit on the input range
+ACTOR Future<Void> auditInputRangeOnAllStorageServers(Reference<DataDistributor> self,
+                                                      std::shared_ptr<DDAudit> audit,
+                                                      KeyRange range) {
+	ASSERT(audit->coreState.getType() == AuditType::ValidateStorageServerShard);
+	TraceEvent(SevInfo, "DDAuditInputRangeOnAllStorageServersBegin", self->ddId)
+	    .detail("AuditID", audit->coreState.id)
+	    .detail("AuditType", audit->coreState.getType());
+	state StorageServerInterface targetServer;
+	try {
+		state ServerWorkerInfos serverWorkers = wait(self->txnProcessor->getServerListAndProcessClasses());
+		state int i = 0;
+		for (; i < serverWorkers.servers.size(); ++i) {
+			targetServer = serverWorkers.servers[i].first;
+			// Currently, Tss server may not follow the auit consistency rule
+			// Thus, skip if the server is tss
+			if (targetServer.isTss()) {
+				continue;
+			}
+			audit->actors.add(
+			    auditMakeProgressOnServer(self, audit, range, targetServer, /*makeProgressByServer=*/true));
+			wait(delay(0.1));
 		}
+	} catch (Error& e) {
+		TraceEvent(SevWarn, "DDAuditInputRangeOnAllStorageServersError", self->ddId)
+		    .errorUnsuppressed(e)
+		    .detail("AuditID", audit->coreState.id)
+		    .detail("AuditType", audit->coreState.getType());
+		throw e;
 	}
-	self->audits[auditState.getType()].erase(auditState.id);
+	TraceEvent(SevInfo, "DDAuditInputRangeOnAllStorageServersEnd", self->ddId)
+	    .detail("AuditID", audit->coreState.id)
+	    .detail("AuditType", audit->coreState.getType());
 
 	return Void();
 }
 
-ACTOR Future<Void> loadAndDispatchAuditRange(Reference<DataDistributor> self,
+// Schedule audit task on the input storage server (ssi)
+// Option makeProgressbyServer:
+// If we store the progress of complete range for each individual server,
+// we should set makeProgressbyServer == true. Then, we load the progress on each server
+// If we store the progress of complete range without distinguishing servers,
+// we should set makeProgressbyServer == false. Then, we load the progress globally
```

Comment:
nit: `makeProgressbyServer` seems to be redundant to the AuditType?

Reply:
We want auditMakeProgressOnServer to be decoupled with a particular auditType for the simplicity. Thus, we provide this option to configure the function to behave as required for any auditType that uses this function.